python/dgl/transforms/functional.py

```diff
@@ -1828,20 +1838,64 @@ def add_self_loop(g, etype=None):
     ...                                   torch.tensor([0, 1])),
     ...     ('user', 'plays', 'game'): (torch.tensor([0, 1]),
     ...                                 torch.tensor([0, 1]))})
-    >>> g = dgl.add_self_loop(g, etype='follows')
+    >>> g.edata['feat'] = {('user', 'follows', 'user'): torch.randn(2, 5),
+    ...                    ('user', 'plays', 'game'): torch.randn(2, 5)}
+    >>> g.edata['feat1'] = {('user', 'follows', 'user'): torch.randn(2, 15),
+    ...                     ('user', 'plays', 'game'): torch.randn(2, 15)}
+    >>> g = dgl.add_self_loop(g, edge_feat_names=['feat', 'feat1'], fill_data='mean',
+    ...                       etype='follows')
     >>> g
     Graph(num_nodes={'user': 3, 'game': 2},
-        num_edges={('user', 'plays', 'game'): 2, ('user', 'follows', 'user'): 5},
-        metagraph=[('user', 'user'), ('user', 'game')])
+          num_edges={('user', 'plays', 'game'): 2, ('user', 'follows', 'user'): 5},
+          metagraph=[('user', 'user', 'follows'), ('user', 'game', 'plays')])
+    >>> g.edata['feat']
+    {('user', 'follows', 'user'): tensor([[-0.7737, -1.4461, -0.2310, -1.2934, -0.3273],
+                                          [ 0.7572, -0.0286,  0.6663, -0.1792,  1.6792],
+                                          [-0.7737, -1.4461, -0.2310, -1.2934, -0.3273],
+                                          [ 0.7572, -0.0286,  0.6663, -0.1792,  1.6792],
+                                          [ 0.0000,  0.0000,  0.0000,  0.0000,  0.0000]]),
+     ('user', 'plays', 'game'): tensor([[ 0.3663, 0.1767, 0.1413, 0.2587, -0.8130],
+                                        [-1.0883,  0.2876, -0.6633,  0.7246,  1.3131]])}
     """
     etype = g.to_canonical_etype(etype)
+    data = {}
+    reduce_funcs = {'sum': function.sum,
+                    'mean': function.mean,
+                    'max': function.max,
+                    'min': function.min}
+
+    if edge_feat_names is None:
+        edge_feat_names = []
+
     if etype[0] != etype[2]:
         raise DGLError(
             'add_self_loop does not support unidirectional bipartite graphs: {}.' \
             'Please make sure the types of head node and tail node are identical.' \
             ''.format(etype))
+
+    for feat_name in edge_feat_names:
+        if isinstance(fill_data, (int, float)):
+            dtype = g.edges[etype[1]].data[feat_name].dtype
```

Comment:
L1878 does not work if there are multiple canonical edge types with the same middle edge type name.